ocdata/core/slab.py

```diff
@@ -0,0 +1,529 @@
+import math
+import os
+import pickle
+from collections import defaultdict
+
+import ase
+import numpy as np
+import pymatgen
+from ase import neighborlist
+from ase.constraints import FixAtoms
+from pymatgen.analysis.local_env import VoronoiNN
+from pymatgen.core import Composition, Structure
+from pymatgen.core.surface import (
+    SlabGenerator,
+    get_symmetrically_distinct_miller_indices,
+)
+from pymatgen.io.ase import AseAtomsAdaptor
+from pymatgen.symmetry.analyzer import SpacegroupAnalyzer
+
+from ocdata.configs.constants import MIN_XY
+
+
+class Slab:
+    """
+    Initializes a slab object, i.e. a particular slab tiled along xyz, in
+    one of 2 ways:
+    - Pass in a Bulk object and a slab 4-tuple containing
+    (atoms, miller, shift, top).
+    - Pass in a Bulk object and randomly sample a slab.
+
+    Arguments
+    ---------
+    bulk: Bulk
+        Corresponding Bulk object.
+    unit_slab_struct: Structure
+        Unit cell slab structure.
+    millers: tuple
+        Miller indices of slab.
+    shift: float
+        Shift of slab.
+    top: bool
+        Whether slab is top or bottom.
+    tile_and_tag: bool
+        Whether to tile slab along xyz and tag surface / fixed atoms.
+    """
+
+    def __init__(
+        self,
+        bulk=None,
+        unit_slab_struct: Structure = None,
+        millers: tuple = None,
+        shift: float = None,
+        top: bool = None,
+        tile_and_tag: bool = True,
+    ):
+        assert bulk is not None
+        self.bulk = bulk
+
+        self.atoms = AseAtomsAdaptor.get_atoms(unit_slab_struct)
+        self.millers = millers
+        self.shift = shift
+        self.top = top
+
+        if tile_and_tag:
+            self.atoms = tile_atoms(self.atoms)
+            self.atoms = tag_surface_atoms(self.atoms, self.bulk.atoms)
+            self.atoms = set_fixed_atom_constraints(self.atoms)
+
+        assert (
+            Composition(self.atoms.get_chemical_formula()).reduced_formula
+            == Composition(bulk.atoms.get_chemical_formula()).reduced_formula
+        ), "Mismatched bulk and surface"
+
+    @classmethod
+    def from_bulk_get_random_slab(cls, bulk=None, max_miller=2, tile_and_tag=True):
+        assert bulk is not None
+
+        slabs = compute_slabs(
+            bulk.atoms,
+            max_miller=max_miller,
+        )
+        slab_idx = np.random.randint(len(slabs))
+        unit_slab_struct, millers, shift, top = slabs[slab_idx]
+        return cls(bulk, unit_slab_struct, millers, shift, top, tile_and_tag)
+
+    @classmethod
+    def from_bulk_get_all_slabs(cls, bulk=None, max_miller=2, tile_and_tag=True):
```

Comment:
Does it make sense to have this in the Slab class? The more typical use case is that given one bulk, we want to generate inputs for many slabs, so we wouldn't want to recalculate this every slab. Instead we should calculate it once per bulk and store the list of slabs in the bulk class

Reply:
> The more typical use case is that given one bulk, we want to generate inputs for many slabs, so we wouldn't want to recalculate this every slab.

That usecase _is_ the primary intent, something like:
```
bulk = Bulk(bulk_id_from_db=0)
slabs = bulk.get_slabs()
```
Internally, `bulk.get_slabs` calls the `@classmethod` from `Slab`. Having all ways of initializing slabs in one place helps with managing the code / readability. But let me know if you feel strongly and have a mock spec of what you have in mind.

Reply:
Discussed offline, this is fine, and we'll add functionality for downstream runner scripts to generate inputs for all surfaces given one bulk and a list of adsorbates